### DISCUSSION TITLE
Flatpak package for linux

### DISCUSSION THREAD

**Discussion by aryan02420** [efroemling/ballistica#459] [General]
Hi Eric, I was going through the wiki and noticed the section about [Tool/Pipeline Improvements][wiki]

I would like to help you distribute BombSquad as [Flatpak][flatpak] package for linux via [Flathub][flathub]. Flatpak is integrated into many Linux App stores by default, which might boost the discover-ability of bombsquad.

The bombsquad executable will be pulled directly from [ballistica.net][builds] and bundled with its dependencies, so no changes are needed in this repo :)

[wiki]: https://github.com/efroemling/ballistica/wiki/Contributing#toolpipeline-improvements
[flatpak]: https://docs.flatpak.org/en/latest/
[flathub]: https://flathub.org/home
[builds]: https://tools.ballistica.net/builds

**Comment by Dliwk** [comment]
Oh cool; I didn't see anyone using Flatpak here before :))

Actually, I have my own flatpak manifest for BombSquad and I'd love to share it here if it would help (though I'm not at home right now).

But it has some visual glitch (it started when I updated to Fedora 36 which also forced me to use nouveau instead of proprietary nvidia driver):
![изображение](https://user-images.githubusercontent.com/47295209/177827597-dfaf581d-05b1-408f-a49d-336c9c0e3d6e.png)
(this appears only with flatpak version of game)

Which distro are you running? Or maybe anyone else has seen this too?

**Reply by efroemling** [reply]
Weird. I'd be curious if things behave better if you set graphics settings to medium or low. But I haven't seen things misbehave in that exact way before.. (especially the blue tinted buttons)

> But it has some visual glitch (it started when I updated to Fedora 36 which also forced me to use nouveau instead of proprietary nvidia driver): ![изображение](https://user-images.githubusercontent.com/47295209/177827597-dfaf581d-05b1-408f-a49d-336c9c0e3d6e.png)

**Reply by aryan02420** [reply]
I am using openSUSE Tumbleweed, Intel UHD 620

**Reply by Dliwk** [reply]
> Weird. I'd be curious if things behave better if you set graphics settings to medium or low. But I haven't seen things misbehave in that exact way before.. (especially the blue tinted buttons)

Haha oops this blue outline happened because of one of my plugin, but exactly the same problem exists even without any mod. :grimacing:

**Reply by Dliwk** [reply]
Only wanted to mention that I finally got home and installed back nvidia drivers and everything became normal.

**Comment by efroemling** [comment]
> 

That sounds great; thanks!

One thought:
The test builds on the ballistica.net site might not always be in a fully stable state and will always show 'Test Build' in the menu. If by chance you'd like to add a module under tools/batools or something that creates the Flatpak I'd be happy to set that up to run whenever I push out official stable builds. Though if you prefer to just keep things on your end and pull from the ballistica.net test builds that's fine too. 👍

**Reply by Dliwk** [reply]
I'm not an expert, but flatpak packages often use `extra-data` field, which points to some static data, which downloaded by the client separately from the package itself, reducing package size and perhaps load on the repository with packages. So basically storing somewhere 'stable' build for Flatpak, without 'Test Build' in the menu would be fine.

**Reply by aryan02420** [reply]
@efroemling Great!
Do you have any preference for the application ID? `net.froemling.bombsquad`, `net.ballistica.bombsquad` or something else?

Also what should be the license of the final package? Can I label it as Proprietary and link to [LICENSE](https://github.com/efroemling/ballistica/blob/master/LICENSE)

> Custom licenses which are not in the SPDX registry, like proprietary licenses, can be denoted using the LicenseRef notation. LicenseRef-proprietary can be used to denote a proprietary license, with an optional URL to the license text following after a = sign.
>
> The license given in the project_license tag should be the ‘main’ license of the project. For a software project, this is typically the license for the code.

flatpak supports branches, so we can have `stable` and `latest` branches. You can either create a webhook to notify the flatpak repository about a new stable release or we can just use an older build from https://files.ballistica.net/bombsquad/builds/old/ as the stable version.

For webhooks, we can set up an action with `on: release: types: ['publish']` trigger or configure one from github.com/efroemling/ballistica/settings/hooks/new. I prefer using an older build (for now) since it requires almost no change to either of our setup

**Reply by aryan02420** [reply]
@Dliwk 
There is an `x-checker-data` field to find if newer versions of resource exist.
Each downloaded file is cached and requires its hash to match the one provided in the manifest file. New versions will only be downloaded when the hash changes. 
`x-data-checker` can be used to specify a version pattern, all matching patterns will be recorded and the latest version will be downloaded and reported to the maintainer. This can work nicely with Github Actions.

I am not sure if different branches share their cache. If not, then it is only one extra download/stable release

**Reply by Dliwk** [reply]
Hi there, I finally got home :partying_face: 

@aryan02420 Here is what I use: https://gist.github.com/Dliwk/eaec7766e32859a86d096591ea12bc41 This is pretty simple manifest generator but hope it may help.

**Comment by Tomin1** [comment]
Hi! Has there been any progress on having a flatpak build? I tried to run BombSquad on my laptop but it wants 32-bit libpython and I really don't want to enable multilib on this machine. A flatpak or appimage would make installing easy without cluttering my system with dependencies. (Also makes me wonder why a release with x86_64 in its name is built with 32-bit library.)

**Reply by EraOSBeta** [reply]
I'm not aware of any developments on a Flatpak build, but there is a Docker image [here](https://github.com/efroemling/ballistica/pkgs/container/ballistica%2Fbombsquad_gui) and [there](https://codeberg.org/3ra/-/packages/container/ballisticakit/1.7.39-gui-debug) if you're interested; only problem with the Docker builds is that you can't play tournaments with them if I remember correctly.
Also as far as i know the builds don't require a 32-bit python library, they require a python-dev library (specifically `python3.12-dev` for the latest version of the game) which are easily accessible for some distros (eg. on some Debian-based distros you can do `sudo apt install python3.12-dev` or use the [deadsnakes PPA](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa) for older Ubuntu-based distros).
And if the previous two options don't work for you you could try playing the Windows version through [Wine](https://www.winehq.org/).

**Reply by Tomin1** [reply]
Hmm. Well, Arch Linux doesn't have Python 3.12 at all. It is in AUR but I'd rather not install it... Oh, well.

I should try the docker build though, thanks for the tip. Tournaments are not something I'd need anyway. The idea was to able to play with a friend locally. (I tried, it can't find audio devices and that seems quite problematic overall. This probably won't help me right now.)

**Reply by Loup-Garou911XD** [reply]
> Hmm. Well, Arch Linux doesn't have Python 3.12 at all. It is in AUR but I'd rather not install it... Oh, well.
> 
> I should try the docker build though, thanks for the tip. Tournaments are not something I'd need anyway. The idea was to able to play with a friend locally. (I tried, it can't find audio devices and that seems quite problematic overall. This probably won't help me right now.)

Well the game and arch linux system python version are same as of now
Also there is a wiki page for docker builds if that helps

**Comment by Loup-Garou911XD** [comment]
update
i did some work and got this to build on github action in reasonable time with caching
#855

**Reply by Loup-Garou911XD** [reply]
thanks to @Dliwk 's script for getting me started

**Reply by Loup-Garou911XD** [reply]
> update i did some work and got this to build on github action in reasonable time with caching #855

albeit with 36 commits and rigorous trial and error
